### PR TITLE
Hare programming language: init

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lourkeur** [comment]
Might be made redundant by #170569, can you take a look?

**Comment by AndersonTorres** [comment]
Summarizing: I did not like the code he wrote.

File: pkgs/development/compilers/hare/harec.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
+{ lib
+, stdenv
+, fetchFromSourcehut
+, qbe
+, scdoc
+}:
+
+stdenv.mkDerivation rec {
+  pname = "harec";
+  version = "0.pre+date=2022-04-25";
```
Comment (AtilaSaraiva) [inline pkgs/development/compilers/hare/harec.nix:10]: Is this kind of versioning allowed? Wouldn't be better to use the unstable format?

**Review by AtilaSaraiva — COMMENTED** [review]
Very nice, i'm eager to see the final product.

File: pkgs/development/compilers/hare/harec.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/development/compilers/hare/harec.nix:10]: https://github.com/NixOS/rfcs/pull/107

Comment (AndersonTorres) [reply pkgs/development/compilers/hare/harec.nix:10]: (OTOH I have not so much hope...)

**Comment by siraben** [comment]
Can this not be marked as broken on darwin? I understand upstream's sentiment but it _is_ free software and we have the right to package it for as many operating systems as we want.

**Comment by alyssais** [comment]
Does it _build_ for Darwin?

**Comment by siraben** [comment]
@alyssais qbe fails to build so this fails transitively.

**Comment by AndersonTorres** [comment]
> Can this not be marked as broken on darwin? 

It can, but it will be useless...

> I understand upstream's sentiment but it _is_ free software and we have the right to package it for as many operating systems as we want.

Okay, go for it.

[ ] Port QBE to Darwin
[ ] Port the compiler to Darwin
[ ] Port the runtime to Darwin
[ ] Debug and iterate
[ ] Maintain that port (or the patchset) in a separate tree because the upstream will refuse to provide support for a proprietary platform (well, Darwin kernel is open source but the only "distro" really running it is the Cupertinian one :) )
[ ] Provide the patches to this Nix expression

Follow the model:

https://briancallahan.net/blog/20220427.html

**Comment by siraben** [comment]
@AndersonTorres I understand that the work is non-trivial to maintain on darwin, I guess I was being idealistic with the suggestion 😂

**Review by kamronm — COMMENTED** [review]
Tested a few examples from the hare tutorial on both x86 and aarch64 NixOS and it worked well.

Everything looks good to me, but it would be nice for people with more experience to take a look given the two pull requests for hare (e.g. do dependencies need to be specified in both buildInputs and nativeBuildInputs? is this what depsBuildTarget is for? is it even necessary given the other PR uses only one or the other?)

**Comment by AndersonTorres** [comment]
> do dependencies need to be specified in both buildInputs and nativeBuildInputs? 

`strictDeps = true;` forces this sometimes.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 170359` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>harec</li>
  </ul>
</details>

**Comment by AndersonTorres** [comment]
As a "bug report", I should... report that most recent Hare drivers are not working: they stuck indefinitely trying to recompile harec.

File: pkgs/development/compilers/hare/harec.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  meta = with lib; {
+    homepage = "http://harelang.org/";
+    description = "Bootstrapping Hare compiler written in C for POSIX systems";
+    license = licenses.gpl3Only;
+    maintainers = with maintainers; [ AndersonTorres ];
+    platforms = with platforms; unix;
+    # The upstream developers do not like proprietary operating systems;
+    # therefore this should be "broken" in Darwin for almost ever...
+    broken = stdenv.isDarwin;
```
Comment (alyssais) [inline pkgs/development/compilers/hare/harec.nix:42]: This isn't broken, it's unsupported.  So `meta.platforms` should be refined from "unix" to the specific platforms upstream supports:

```nix
intersectLists (platforms.freebsd ++ platforms.linux) (platforms.aarch64 ++ platforms.x86_64 ++ platforms.riscv64)
```

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 170359` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hare</li>
    <li>harec</li>
  </ul>
</details>

**Comment by AndersonTorres** [comment]
> Author of #170569, all fair and I am more than happy to abandon that PR. It would have been nice to receive a poke though so that I could abandon it rather than notice this PR as a rebase conflict post-merge. '^^

Sorry by the terse response. 
My expression was almost complete, however I did not upload it because of real life issues.
Would be a waste of time to comment another code and end by copy-pasting my code above another one.

> 
> Overall, your expression is better and I think you are more competent with Nix than I am.

Many thanks for the compliments! But I am a novice, I have so much to learn.

> The test suite is only running for `harec` though and not `hare`, will send a PR. 

Initially I have not seen a `make test` on hare tree, only in `harec`. Thanks for notice this!

> Also, note that `hare release` relies on OpenSSH and Git being in `PATH`, whether we want to wrap them I am not sure as they are fairly big and not essential for most Hare usage.

As you have noticed, and also as I have read in [this blog post](https://harelang.org/blog/2021-11-22-hare-release/), `hare release` is completely optional, has two big dependencies **and** is strongly tied to git (while the author welcomes future VCSes). I was thinking in create a third entry point, something like `hare-full` containing these extras **and** such that it does not trigger an unnecessary rebuild.

However it would require a bit more complicated procedure. I can do that later...

> I am also not sure why you opt for maintaining a copy `config.mk` rather than rely on `makeFlags`

I prefer to follow/emulate most closely the upstream approach when including things to Nixpkgs. This way I can be sure that the build is going as they wanted.

The upstream project [suggests](https://harelang.org/installation/) to create a `config.mk` file and edit/populate it according to the local environment.
So I preferred this approach of "write the file" because it emulates that original intention. 
In other distros you would do this type of thing with a bunch of sed&awk, but in Nixpkgs you can use those utilities to write files on demand!

Indeed I have done something like this in [openmsx](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/emulators/openmsx/default.nix) before.

**Pull request merged**